internal/provider/resource_storage_bucket.go

```diff
@@ -0,0 +1,553 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package provider
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"strconv"
+
+	"github.com/hashicorp/boundary/api"
+	"github.com/hashicorp/boundary/api/storagebuckets"
+	"github.com/hashicorp/go-secure-stdlib/parseutil"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+const (
+	storageBucketNameKey           = "bucket_name"
+	storageBucketPrefixKey         = "bucket_prefix"
+	defaultStorageBucketPluginName = "aws"
+)
+
+func resourceStorageBucket() *schema.Resource {
+	return &schema.Resource{
+		Description:   "The storage bucket resource allows you to configure a Boundary storage bucket. A storage bucket can only belong to the Global scope or an Org scope.",
+		CreateContext: resourceStorageBucketCreate,
+		ReadContext:   resourceStorageBucketRead,
+		UpdateContext: resourceStorageBucketUpdate,
+		DeleteContext: resourceStorageBucketDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			IDKey: {
+				Description: "The ID of the storage bucket.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			NameKey: {
+				Description: "The storage bucket name. Defaults to the resource name.",
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+			DescriptionKey: {
+				Description: "The storage bucket description.",
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+			PluginIdKey: {
+				Description:   "The ID of the plugin that should back the resource. This or " + PluginNameKey + " must be defined.",
+				Type:          schema.TypeString,
+				ConflictsWith: []string{PluginNameKey},
+				ExactlyOneOf:  []string{PluginIdKey, PluginNameKey},
+				Optional:      true,
+				ForceNew:      true,
+				Computed:      true, // If name is provided this will be computed
+			},
+			PluginNameKey: {
+				Description:   "The name of the plugin that should back the resource. This or " + PluginIdKey + " must be defined.",
+				Type:          schema.TypeString,
+				ConflictsWith: []string{PluginIdKey},
+				ExactlyOneOf:  []string{PluginIdKey, PluginNameKey},
+				Default:       defaultStorageBucketPluginName,
```

Comment:
I don't think there should be a default here. 
Apart from that won't this having a default actually break the behavior of providing either the Id or name? If the user provided the `id` but no `name` it would default to `aws`, would it not then return an error because both were provided?

Reply:
That's a good point. I didn't think of that. I added a change to set the default plugin name during creation. Maybe we're better off not setting the default value. The user will have to set it 🤔

Reply:
yeah I think there should be no default type and always have the user set it.